File: scripts/generate_audio_tests.py
#!/usr/bin/env python3
"""
Generate audio test files for end-to-end audio validation.

Generates WAV and/or FLAC in multiple channel counts, sample rates and bit depths.
Signals:
- id: per-channel identification tones (unique frequency per channel)
- thd1k: single-tone 1 kHz @ -3 dBFS (for THD)
- thd100: single-tone 100 Hz @ -3 dBFS (low-frequency THD)
- imd_smpte: SMPTE two-tone 60 Hz + 7 kHz (4:1 amplitude ratio)
- imd_ccif: CCIF two-tone 19 kHz + 20 kHz (equal amplitudes)

Examples:
  python3 scripts/generate_audio_tests.py --out-dir target/test-audio \
    --formats wav flac --channels 1 2 6 16 --sample-rates 44100 48000 96000 192000 --bits 16 24 \
    --signals id thd1k imd_ccif --duration 3

Notes:
- FLAC writing prefers the Python soundfile module; if unavailable, falls back to the `flac` CLI.
- WAV writer uses the Python stdlib `wave` module and supports 16/24-bit PCM.
- Cross-platform (macOS/Linux/Windows) if Python and dependencies are available.
"""

from __future__ import annotations
import argparse
import math
import os
import sys
import json
import shutil
from pathlib import Path
from typing import List, Tuple, Optional

import numpy as np

try:
    import soundfile as sf  # optional, used for FLAC
    HAS_SF = True
except Exception:
    HAS_SF = False


def db_to_linear(db: float) -> float:
    return 10.0 ** (db / 20.0)


def tone(freq: float, sr: int, dur: float, phase: float = 0.0, amp: float = 0.707) -> np.ndarray:
    t = np.arange(int(sr * dur), dtype=np.float64) / sr
    return (amp * np.sin(2.0 * np.pi * freq * t + phase)).astype(np.float32)


def two_tone(freq1: float, freq2: float, sr: int, dur: float, amp1: float, amp2: float) -> np.ndarray:
    x1 = tone(freq1, sr, dur, amp=amp1)
    x2 = tone(freq2, sr, dur, amp=amp2)
    x = x1 + x2
    peak = np.max(np.abs(x))
    if peak > 0.999:
        x = (x / peak) * 0.999  # keep headroom
    return x.astype(np.float32)


def per_channel_id_signal(channels: int, sr: int, dur: float) -> np.ndarray:
    """
    Build a multi-channel signal where each channel is a unique single-tone.
    Frequencies are spaced to avoid overlap below 6 kHz and within Nyquist at 44.1 kHz.
    """
    # base freq 300 Hz, step 300 Hz => 300, 600, 900, ...
    # cap within 6 kHz
    freqs = [min(300.0 + 300.0 * i, 6000.0) for i in range(channels)]
    data = [tone(freqs[i], sr, dur, amp=0.707) for i in range(channels)]
    return np.stack(data, axis=1), freqs


def thd_signal(freq: float, channels: int, sr: int, dur: float) -> np.ndarray:
    sig = tone(freq, sr, dur, amp=0.707)
    data = [sig.copy() for _ in range(channels)]
    return np.stack(data, axis=1)


def imd_smpte_signal(channels: int, sr: int, dur: float) -> np.ndarray:
    # SMPTE: 60 Hz at higher amplitude, 7 kHz at lower amplitude (typical 4:1 ratio)
    low = tone(60.0, sr, dur, amp=0.8)  # ~ -1.9 dBFS
    high = tone(7000.0, sr, dur, amp=0.2)  # lower level to avoid clipping
    sig = low + high
    peak = np.max(np.abs(sig))
    if peak > 0.999:
        sig = (sig / peak) * 0.999
    data = [sig.astype(np.float32) for _ in range(channels)]
    return np.stack(data, axis=1)


def imd_ccif_signal(channels: int, sr: int, dur: float) -> np.ndarray:
    # CCIF: equal amplitudes at 19 kHz and 20 kHz
    sig = two_tone(19000.0, 20000.0, sr, dur, amp1=0.5, amp2=0.5)
    data = [sig for _ in range(channels)]
    return np.stack(data, axis=1)


def float_to_int_pcm(x: np.ndarray, bits: int) -> bytes:
    assert x.dtype == np.float32
    # clip
    x = np.clip(x, -0.999999, 0.999999)
    if bits == 16:
        y = (x * 32767.0).round().astype('<i2')
        return y.tobytes()
    elif bits == 24:
        # 24-bit little-endian in 3 bytes
        y = (x * 8388607.0).round().astype(np.int32)
        # pack 3 LSB bytes per sample
        b = bytearray()
        for v in y:
            b.append(v & 0xFF)
            b.append((v >> 8) & 0xFF)
            b.append((v >> 16) & 0xFF)
        return bytes(b)
    else:
        raise ValueError(f"Unsupported PCM bit depth: {bits}")


def write_wav(path: Path, data: np.ndarray, sr: int, bits: int) -> None:
    import wave
    n_channels = data.shape[1]
    n_frames = data.shape[0]

    # interleave
    interleaved = data.reshape(n_frames, n_channels).astype(np.float32)
    interleaved = interleaved  # already interleaved by axis=1 in stacking
    # convert to PCM
    pcm_bytes = b''.join(float_to_int_pcm(interleaved[:, ch], bits) for ch in range(n_channels))
    # But above concatenates channel-by-channel; need frame interleaving
    # Rebuild with per-frame interleave to ensure proper ordering
    # Build per-frame bytes
    bytes_per_sample = 2 if bits == 16 else 3
    frame_bytes = bytearray()
    for i in range(n_frames):
        for ch in range(n_channels):
            sample = float_to_int_pcm(interleaved[i:i+1, ch], bits)
            frame_bytes.extend(sample)

    with wave.open(str(path), 'wb') as wf:
        wf.setnchannels(n_channels)
        wf.setsampwidth(2 if bits == 16 else 3)
        wf.setframerate(sr)
        wf.writeframes(frame_bytes)


def write_flac(path: Path, data: np.ndarray, sr: int, bits: int) -> None:
    global HAS_SF
    if HAS_SF:
        subtype = 'PCM_24' if bits == 24 else 'PCM_16'
        sf.write(str(path), data, sr, subtype=subtype, format='FLAC')
        return
    # fallback to `flac` CLI using a temp WAV
    if shutil.which('flac') is None:
        raise RuntimeError("FLAC output requested but neither soundfile module nor 'flac' CLI is available")
    import tempfile
    with tempfile.TemporaryDirectory() as td:
        tmp_wav = Path(td) / 'tmp.wav'
        write_wav(tmp_wav, data, sr, bits)
        # -f overwrite, -s silent
        cmd = ['flac', '-f', '-s', str(tmp_wav), '-o', str(path)]
        rc = os.spawnvp(os.P_WAIT, cmd[0], cmd)
        if rc != 0:
            raise RuntimeError(f"flac encoding failed with code {rc}")


def ensure_dir(p: Path) -> None:
    p.mkdir(parents=True, exist_ok=True)


def build_filename(signal: str, channels: int, sr: int, bits: int, ext: str) -> str:
    return f"{signal}_ch{channels}_sr{sr}_b{bits}.{ext}"


def generate_one(out_dir: Path, fmt: str, signal: str, channels: int, sr: int, bits: int, dur: float) -> Path:
    if channels < 1 or channels > 16:
        raise ValueError("channels must be between 1 and 16")
    if bits not in (16, 24):
        raise ValueError("bits must be 16 or 24")

    # Make signal
    if signal == 'id':
        data, freqs = per_channel_id_signal(channels, sr, dur)
        manifest = {"type": "id", "freqs": freqs}
    elif signal == 'thd1k':
        data = thd_signal(1000.0, channels, sr, dur)
        manifest = {"type": "thd1k", "freq": 1000.0}
    elif signal == 'thd100':
        data = thd_signal(100.0, channels, sr, dur)
        manifest = {"type": "thd100", "freq": 100.0}
    elif signal == 'imd_smpte':
        data = imd_smpte_signal(channels, sr, dur)
        manifest = {"type": "imd_smpte", "freqs": [60.0, 7000.0], "ratio": "~4:1"}
    elif signal == 'imd_ccif':
        data = imd_ccif_signal(channels, sr, dur)
        manifest = {"type": "imd_ccif", "freqs": [19000.0, 20000.0]}
    else:
        raise ValueError(f"unknown signal: {signal}")

    # Make directories by format/signal
    subdir = out_dir / fmt / signal
    ensure_dir(subdir)
    fname = build_filename(signal, channels, sr, bits, fmt)
    out_path = subdir / fname

    # Write
    if fmt == 'wav':
        write_wav(out_path, data, sr, bits)
    elif fmt == 'flac':
        write_flac(out_path, data, sr, bits)
    else:
        raise ValueError("format must be wav or flac")

    # Write sidecar JSON with metadata
    sidecar = out_path.with_suffix(out_path.suffix + '.json')
    with open(sidecar, 'w') as f:
        json.dump({
            "format": fmt,
            "channels": channels,
            "sample_rate": sr,
            "bits": bits,
            "duration": dur,
            "signal": manifest,
        }, f, indent=2)

    return out_path


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description="Generate audio test files")
    p.add_argument('--out-dir', type=Path, default=Path('target/test-audio'), help='Output directory')
    p.add_argument('--formats', nargs='+', default=['wav', 'flac'], choices=['wav', 'flac'])
    p.add_argument('--channels', nargs='+', type=int, default=[2, 6, 16])
    p.add_argument('--sample-rates', nargs='+', type=int, default=[44100, 96000, 192000])
    p.add_argument('--bits', nargs='+', type=int, default=[16, 24])
    p.add_argument('--signals', nargs='+', default=['id', 'thd1k', 'thd100', 'imd_smpte', 'imd_ccif'],
                   choices=['id', 'thd1k', 'thd100', 'imd_smpte', 'imd_ccif'])
    p.add_argument('--duration', type=float, default=3.0, help='Duration in seconds')
    return p.parse_args()


def main() -> int:
    args = parse_args()
    ensure_dir(args.out_dir)

    generated = []
    for fmt in args.formats:
        for ch in args.channels:
            for sr in args.sample_rates:
                nyq = sr / 2.0
                for bits in args.bits:
                    for sig in args.signals:
                        # avoid CCIF at too-low sample rates if tones are too close to Nyquist
                        if sig == 'imd_ccif' and nyq < 20500.0:
                            # Still acceptable at 44.1k (nyq 22050), so keep it; skip only if nyq < 20500
                            pass
                        try:
                            path = generate_one(args.out_dir, fmt, sig, ch, sr, bits, args.duration)
                            generated.append(str(path))
                        except Exception as e:
                            print(f"[WARN] failed to generate {fmt} {sig} ch{ch} sr{sr} b{bits}: {e}", file=sys.stderr)
                            continue

    manifest_path = args.out_dir / 'manifest.json'
    with open(manifest_path, 'w') as f:
        json.dump({"files": generated}, f, indent=2)
    print(f"Generated {len(generated)} files. Manifest: {manifest_path}")
    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: src-ui-backend/bin/generate_audio_tests.rs
#!/usr/bin/env rust
//! Generate audio test files for end-to-end audio validation.
//!
//! Generates WAV files in multiple channel counts, sample rates and bit depths.
//! Signals:
//! - id: per-channel identification tones (unique frequency per channel)
//! - thd1k: single-tone 1 kHz @ -3 dBFS (for THD)
//! - thd100: single-tone 100 Hz @ -3 dBFS (low-frequency THD)
//! - imd_smpte: SMPTE two-tone 60 Hz + 7 kHz (4:1 amplitude ratio)
//! - imd_ccif: CCIF two-tone 19 kHz + 20 kHz (equal amplitudes)
//! - sweep: logarithmic frequency sweep from 20 Hz to 20 kHz (10s fixed duration)

use clap::{Parser, ValueEnum};
use hound::{SampleFormat, WavSpec, WavWriter};
use serde::{Deserialize, Serialize};
use serde_json;
use std::f32::consts::PI;
use std::fs;
use std::path::{Path, PathBuf};

// Constants
const AMP_STD: f32 = 0.707; // ~-3 dBFS
const SMPTE_AMP1: f32 = 0.8; // 60 Hz amplitude
const SMPTE_AMP2: f32 = 0.2; // 7 kHz amplitude
const CCIF_AMP: f32 = 0.5; // 19/20 kHz equal amplitudes
const ID_BASE_FREQ: f32 = 300.0;
const ID_STEP_FREQ: f32 = 300.0;
const ID_MAX_FREQ: f32 = 6000.0;
const SWEEP_DURATION: f32 = 10.0; // Fixed duration for sweep

#[derive(Debug, Clone, Copy, PartialEq, Eq, ValueEnum)]
enum SignalKind {
    Id,
    Thd1k,
    Thd100,
    ImdSmpte,
    ImdCcif,
    Sweep,
}

impl SignalKind {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Id => "id",
            Self::Thd1k => "thd1k",
            Self::Thd100 => "thd100",
            Self::ImdSmpte => "imd_smpte",
            Self::ImdCcif => "imd_ccif",
            Self::Sweep => "sweep",
        }
    }

    fn all() -> Vec<Self> {
        vec![
            Self::Id,
            Self::Thd1k,
            Self::Thd100,
            Self::ImdSmpte,
            Self::ImdCcif,
            Self::Sweep,
        ]
    }
}

#[derive(Parser)]
#[command(name = "generate_audio_tests")]
#[command(about = "Generate audio test files for validation", long_about = None)]
struct Cli {
    /// Output directory
    #[arg(long, default_value = "target/test-audio")]
    out_dir: PathBuf,

    /// Number of channels (comma-separated)
    #[arg(long, value_delimiter = ',', default_values_t = vec![2, 6, 16])]
    channels: Vec<u16>,

    /// Sample rates in Hz (comma-separated)
    #[arg(long = "sample-rates", value_delimiter = ',', default_values_t = vec![44100, 96000, 192000])]
    sample_rates: Vec<u32>,

    /// Bit depths (comma-separated, 16 or 24 only)
    #[arg(long, value_delimiter = ',', default_values_t = vec![16, 24])]
    bits: Vec<u16>,

    /// Signal types to generate (comma-separated)
    #[arg(long, value_delimiter = ',')]
    signals: Vec<SignalKind>,

    /// Duration in seconds (default 3.0, does not apply to sweep which is fixed at 10s)
    #[arg(long, default_value_t = 3.0)]
    duration: f32,
}

#[derive(Debug, Serialize, Deserialize)]
struct Sidecar {
    format: String,
    channels: u16,
    sample_rate: u32,
    bits: u16,
    duration: f32,
    signal: SignalMetadata,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
#[serde(rename_all = "snake_case")]
enum SignalMetadata {
    Id { freqs: Vec<f32> },
    Thd1k { freq: f32 },
    Thd100 { freq: f32 },
    ImdSmpte { freqs: [f32; 2], ratio: u8 },
    ImdCcif { freqs: [f32; 2] },
    Sweep { freq_start: f32, freq_end: f32, kind: String },
}

#[derive(Debug)]
struct GenerationStats {
    generated: usize,
    skipped: usize,
    failed: usize,
}

impl GenerationStats {
    fn new() -> Self {
        Self {
            generated: 0,
            skipped: 0,
            failed: 0,
        }
    }
}

fn main() {
    let cli = Cli::parse();

    // Validate bit depths
    for &bits in &cli.bits {
        if bits != 16 && bits != 24 {
            eprintln!("Error: Bit depth must be 16 or 24, got {}", bits);
            std::process::exit(1);
        }
    }

    // If no signals specified, use all
    let signals = if cli.signals.is_empty() {
        SignalKind::all()
    } else {
        cli.signals.clone()
    };

    // Create output directory
    if let Err(e) = fs::create_dir_all(&cli.out_dir) {
        eprintln!("Error: Failed to create output directory: {}", e);
        std::process::exit(1);
    }

    let mut stats = GenerationStats::new();
    let mut manifest_files = Vec::new();

    // Generate all combinations
    for signal in &signals {
        for &channels in &cli.channels {
            if channels < 1 || channels > 16 {
                eprintln!("Warning: Channel count {} out of range [1,16], skipping", channels);
                stats.skipped += 1;
                continue;
            }

            for &sr in &cli.sample_rates {
                for &bits in &cli.bits {
                    let duration = if *signal == SignalKind::Sweep {
                        SWEEP_DURATION
                    } else {
                        cli.duration
                    };

                    match generate_one(
                        &cli.out_dir,
                        *signal,
                        channels,
                        sr,
                        bits,
                        duration,
                    ) {
                        Ok(path) => {
                            manifest_files.push(path.to_string_lossy().to_string());
                            stats.generated += 1;
                        }
                        Err(e) => {
                            if e.contains("Nyquist") || e.contains("skipped") {
                                stats.skipped += 1;
                            } else {
                                eprintln!(
                                    "Warning: Failed to generate {} ch{} sr{} b{}: {}",
                                    signal.as_str(),
                                    channels,
                                    sr,
                                    bits,
                                    e
                                );
                                stats.failed += 1;
                            }
                        }
                    }
                }
            }
        }
    }

    // Write manifest
    let manifest_path = cli.out_dir.join("manifest.json");
    match write_manifest(&manifest_path, &manifest_files) {
        Ok(_) => {
            println!("\nGenerated {} files. Manifest: {}", stats.generated, manifest_path.display());
        }
        Err(e) => {
            eprintln!("Warning: Failed to write manifest: {}", e);
        }
    }

    println!("Summary: Generated: {}, Skipped: {}, Failed: {}", 
             stats.generated, stats.skipped, stats.failed);
}

fn generate_one(
    out_dir: &Path,
    signal: SignalKind,
    channels: u16,
    sr: u32,
    bits: u16,
    duration: f32,
) -> Result<PathBuf, String> {
    let nyquist = sr as f32 / 2.0;

    // Check Nyquist violations
    match signal {
        SignalKind::Thd1k if 1000.0 >= nyquist => {
            return Err(format!("Nyquist violation: 1000 Hz >= {} Hz (skipped)", nyquist));
        }
        SignalKind::Thd100 if 100.0 >= nyquist => {
            return Err(format!("Nyquist violation: 100 Hz >= {} Hz (skipped)", nyquist));
        }
        SignalKind::ImdSmpte if 7000.0 >= nyquist => {
            return Err(format!("Nyquist violation: 7000 Hz >= {} Hz (skipped)", nyquist));
        }
        SignalKind::ImdCcif if 20000.0 >= nyquist => {
            return Err(format!("Nyquist violation: 20 kHz >= {} Hz (skipped)", nyquist));
        }
        SignalKind::Sweep if 20000.0 >= nyquist => {
            return Err(format!("Nyquist violation: sweep end 20 kHz >= {} Hz (skipped)", nyquist));
        }
        SignalKind::Id => {
            let max_id_freq = (ID_BASE_FREQ + ID_STEP_FREQ * (channels as f32 - 1.0)).min(ID_MAX_FREQ);
            if max_id_freq >= nyquist {
                return Err(format!("Nyquist violation: max ID freq {} Hz >= {} Hz (skipped)", max_id_freq, nyquist));
            }
        }
        _ => {}
    }

    // Generate signal data
    let (audio_data, metadata) = match signal {
        SignalKind::Id => {
            let mut freqs = Vec::new();
            let mut per_channel = Vec::new();
            for ch in 0..channels {
                let freq = (ID_BASE_FREQ * (ch as f32 + 1.0)).min(ID_MAX_FREQ);
                freqs.push(freq);
                per_channel.push(gen_tone(freq, AMP_STD, sr, duration));
            }
            let data = interleave_per_channel(&per_channel);
            (data, SignalMetadata::Id { freqs })
        }
        SignalKind::Thd1k => {
            let mono = gen_tone(1000.0, AMP_STD, sr, duration);
            let data = replicate_mono(&mono, channels);
            (data, SignalMetadata::Thd1k { freq: 1000.0 })
        }
        SignalKind::Thd100 => {
            let mono = gen_tone(100.0, AMP_STD, sr, duration);
            let data = replicate_mono(&mono, channels);
            (data, SignalMetadata::Thd100 { freq: 100.0 })
        }
        SignalKind::ImdSmpte => {
            let mono = gen_two_tone(60.0, SMPTE_AMP1, 7000.0, SMPTE_AMP2, sr, duration);
            let data = replicate_mono(&mono, channels);
            (data, SignalMetadata::ImdSmpte { freqs: [60.0, 7000.0], ratio: 4 })
        }
        SignalKind::ImdCcif => {
            let mono = gen_two_tone(19000.0, CCIF_AMP, 20000.0, CCIF_AMP, sr, duration);
            let data = replicate_mono(&mono, channels);
            (data, SignalMetadata::ImdCcif { freqs: [19000.0, 20000.0] })
        }
        SignalKind::Sweep => {
            let mono = gen_log_sweep(20.0, 20000.0, AMP_STD, sr, duration);
            let data = replicate_mono(&mono, channels);
            (data, SignalMetadata::Sweep {
                freq_start: 20.0,
                freq_end: 20000.0,
                kind: "log".to_string(),
            })
        }
    };

    // Create output directory structure
    let subdir = out_dir.join("wav").join(signal.as_str());
    fs::create_dir_all(&subdir).map_err(|e| format!("Failed to create directory: {}", e))?;

    // Build filename
    let filename = format!("{}_ch{}_sr{}_b{}.wav", signal.as_str(), channels, sr, bits);
    let wav_path = subdir.join(&filename);

    // Write WAV file
    write_wav(&wav_path, &audio_data, sr, channels, bits)?;

    // Write sidecar JSON
    let sidecar = Sidecar {
        format: "wav".to_string(),
        channels,
        sample_rate: sr,
        bits,
        duration,
        signal: metadata,
    };

    let sidecar_path = wav_path.with_extension("wav.json");
    write_sidecar(&sidecar_path, &sidecar)?;

    Ok(wav_path)
}

// DSP Helpers

fn clip(x: f32) -> f32 {
    x.clamp(-0.999_999, 0.999_999)
}

fn frames_for(duration: f32, sr: u32) -> usize {
    (duration * sr as f32).round() as usize
}

// Signal Generators

fn gen_tone(freq: f32, amp: f32, sr: u32, dur: f32) -> Vec<f32> {
    let n_frames = frames_for(dur, sr);
    let mut signal = Vec::with_capacity(n_frames);
    let dphi = 2.0 * PI * freq / sr as f32;
    let mut phase: f32 = 0.0;

    for _ in 0..n_frames {
        signal.push(clip(amp * phase.sin()));
        phase += dphi;
        if phase > 2.0 * PI {
            phase -= 2.0 * PI;
        }
    }

    signal
}

fn gen_two_tone(f1: f32, a1: f32, f2: f32, a2: f32, sr: u32, dur: f32) -> Vec<f32> {
    let n_frames = frames_for(dur, sr);
    let mut signal = Vec::with_capacity(n_frames);
    let dphi1 = 2.0 * PI * f1 / sr as f32;
    let dphi2 = 2.0 * PI * f2 / sr as f32;
    let mut phase1: f32 = 0.0;
    let mut phase2: f32 = 0.0;

    for _ in 0..n_frames {
        let sample = a1 * phase1.sin() + a2 * phase2.sin();
        signal.push(clip(sample));
        phase1 += dphi1;
        phase2 += dphi2;
        if phase1 > 2.0 * PI {
            phase1 -= 2.0 * PI;
        }
        if phase2 > 2.0 * PI {
            phase2 -= 2.0 * PI;
        }
    }

    signal
}

fn gen_log_sweep(f_start: f32, f_end: f32, amp: f32, sr: u32, dur: f32) -> Vec<f32> {
    let n_frames = frames_for(dur, sr);
    let mut signal = Vec::with_capacity(n_frames);
    
    let k = (f_end / f_start).ln() / dur;
    let coefficient = 2.0 * PI * f_start / k;

    for n in 0..n_frames {
        let t = n as f32 / sr as f32;
        let phase = coefficient * ((k * t).exp() - 1.0);
        signal.push(clip(amp * phase.sin()));
    }

    signal
}

// Channel operations

fn interleave_per_channel(per_channel: &[Vec<f32>]) -> Vec<f32> {
    let n_channels = per_channel.len();
    if n_channels == 0 {
        return Vec::new();
    }
    let n_frames = per_channel[0].len();
    let mut interleaved = Vec::with_capacity(n_frames * n_channels);

    for frame in 0..n_frames {
        for ch in 0..n_channels {
            interleaved.push(per_channel[ch][frame]);
        }
    }

    interleaved
}

fn replicate_mono(mono: &[f32], channels: u16) -> Vec<f32> {
    let n_frames = mono.len();
    let mut interleaved = Vec::with_capacity(n_frames * channels as usize);

    for &sample in mono {
        for _ in 0..channels {
            interleaved.push(sample);
        }
    }

    interleaved
}

// WAV writing

fn write_wav(
    path: &Path,
    interleaved: &[f32],
    sr: u32,
    channels: u16,
    bits: u16,
) -> Result<(), String> {
    let spec = WavSpec {
        channels,
        sample_rate: sr,
        bits_per_sample: bits,
        sample_format: SampleFormat::Int,
    };

    let mut writer = WavWriter::create(path, spec)
        .map_err(|e| format!("Failed to create WAV writer: {}", e))?;

    match bits {
        16 => {
            for &sample in interleaved {
                let pcm = (clip(sample) * 32767.0).round() as i16;
                writer.write_sample(pcm)
                    .map_err(|e| format!("Failed to write sample: {}", e))?;
            }
        }
        24 => {
            for &sample in interleaved {
                let pcm = (clip(sample) * 8388607.0).round() as i32;
                writer.write_sample(pcm)
                    .map_err(|e| format!("Failed to write sample: {}", e))?;
            }
        }
        _ => return Err(format!("Unsupported bit depth: {}", bits)),
    }

    writer.finalize()
        .map_err(|e| format!("Failed to finalize WAV: {}", e))?;

    Ok(())
}

// JSON writing

fn write_sidecar(path: &Path, sidecar: &Sidecar) -> Result<(), String> {
    let json = serde_json::to_string_pretty(sidecar)
        .map_err(|e| format!("Failed to serialize sidecar: {}", e))?;
    fs::write(path, json)
        .map_err(|e| format!("Failed to write sidecar: {}", e))?;
    Ok(())
}

fn write_manifest(path: &Path, files: &[String]) -> Result<(), String> {
    let manifest = serde_json::json!({ "files": files });
    let json = serde_json::to_string_pretty(&manifest)
        .map_err(|e| format!("Failed to serialize manifest: {}", e))?;
    fs::write(path, json)
        .map_err(|e| format!("Failed to write manifest: {}", e))?;
    Ok(())
}
